records/generator.go

```diff
@@ -244,106 +183,105 @@ func hostToIP4(hostname string) (string, bool) {
+func (rg *RecordGenerator) InsertState(sj state.State, domain string,
+	ns string, listener string, masters []string, spec labels.Func) error {
 
-	// creates a map with slave IP addresses (IPv4)
 	rg.SlaveIPs = make(map[string]string)
 	rg.SRVs = make(rrs)
 	rg.As = make(rrs)
 
-	for _, slave := range sj.Slaves {
-		address, ok := hostToIP4(slave.PID.Host)
-		if ok {
-			a := "slave." + domain + "."
-			rg.insertRR(a, address, "A")
-			srv := net.JoinHostPort(a, slave.PID.Port)
-			rg.insertRR("_slave._tcp."+domain+".", srv, "SRV")
-		} else {
-			logging.VeryVerbose.Printf("string '%q' for slave with id %q is not a valid IP address", address, slave.ID)
-			address = labels.DomainFrag(address, labels.Sep, spec)
-		}
-		rg.SlaveIPs[slave.ID] = address
-	}
+	rg.frameworkRecords(sj, domain, spec)
+	rg.slaveRecords(sj, domain, spec)
+	rg.listenerRecord(listener, ns)
+	rg.masterRecord(domain, masters, sj.Leader)
 
 	// complete crap - refactor me
 	for _, f := range sj.Frameworks {
 		fname := labels.DomainFrag(f.Name, labels.Sep, spec)
-		if address, ok := hostToIP4(f.PID.Host); ok {
-			a := fname + "." + domain + "."
-			rg.insertRR(a, address, "A")
-			srv := net.JoinHostPort(a, f.PID.Port)
-			rg.insertRR("_framework._tcp."+a, srv, "SRV")
-		}
 
+		// insert taks records
+		tail := fname + "." + domain + "."
 		for _, task := range f.Tasks {
 			hostIP, ok := rg.SlaveIPs[task.SlaveID]
+
 			// skip not running or not discoverable tasks
 			if !ok || (task.State != "TASK_RUNNING") {
 				continue
 			}
 
-			tname := spec(task.Name)
-			sid := slaveIDTail(task.SlaveID)
-			tag := hashString(task.ID)
-			tail := fname + "." + domain + "."
+			// context used to build domain names
+			context := struct {
+				TaskName string
+				TaskID   string
+				SlaveID  string
+			}{
+				SlaveID:  slaveIDTail(task.SlaveID),
+				TaskID:   hashString(task.ID),
+				TaskName: spec(task.Name),
+			}
 
-			// A records for task and task-sid
-			arec := tname + "." + tail
+			// insert canonical A records
+			trec := context.TaskName + "-" + context.TaskID + "-" + context.SlaveID + "." + tail
+			arec := context.TaskName + "." + tail
+			containerIP := task.ContainerIP()
 			rg.insertRR(arec, hostIP, "A")
-			trec := tname + "-" + tag + "-" + sid + "." + tail
 			rg.insertRR(trec, hostIP, "A")
-
-			// A records with container IP
-			if containerIP := task.containerIP(); containerIP != "" {
+			if containerIP != "" {
 				rg.insertRR("_container."+arec, containerIP, "A")
 				rg.insertRR("_container."+trec, containerIP, "A")
 			}
 
-			// SRV records
-			if task.Resources.Ports != "" {
-				ports := yankPorts(task.Resources.Ports)
-				for _, port := range ports {
-					srvhost := trec + ":" + port
-					tcp := "_" + tname + "._tcp." + tail
-					udp := "_" + tname + "._udp." + tail
-					rg.insertRR(tcp, srvhost, "SRV")
-					rg.insertRR(udp, srvhost, "SRV")
-				}
-			}
+			// Add RFC 2782 SRV records
+			task.ForEachPort(func(port string) {
```

Comment:
This sort of closure iteration isn't very idiomatic in Go. What prevents us from having a method which simply returns a slice of ports and ranging over that instead?